src/main/java/org/broadinstitute/barclay/argparser/TaggedArgumentParser.java

```diff
@@ -0,0 +1,333 @@
+
+    /**
+     * Delimiter between key-value pairs in the "logical_name,key1=value1,key2=value2" syntax.
+     */
+    private static final String ARGUMENT_KEY_VALUE_PAIR_DELIMITER = ",";
+
+    /**
+     * Separator between keys and values in the "logical_name,key1=value1,key2=value2" syntax.
+     */
+    private static final String ARGUMENT_KEY_VALUE_SEPARATOR = "=";
+
+    /**
+     * Separator used between option name and logical name.
+     */
+    private static final char ARGUMENT_TAG_NAME_SEPARATOR = ':';
+
+    private static final String USAGE = "Argument must of the form argument_name:logical_name(,key=value)*";
+
+    // Map of surrogate keys to Pair(tag_string, raw_argument_value)
+    private Map<String, Pair<String, String>> tagSurrogates = new HashMap<>();
+
+    // Prefixes used by the opt parser for short/long prefixes
+    private static final String SHORT_OPTION_PREFIX = "-";
+    private static final String LONG_OPTION_PREFIX = "--";
+
+    /**
+     * Given an array of raw arguments provided by the user, return an array of args where tagged arguments
+     * have been replaced with curated arguments containing a key to be used by the parser to retrieve the actual
+     * values.
+     * @param args raw arguments as provided by the user
+     * @return curated string of arguments to be presented to the opt parser
+     */
+    public String[] processTaggedOptions(final String[] args) {
+        List<String> finalArgs = new ArrayList<>(args.length);
+        for (int i = 0; i < args.length; ) {
+            String arg = args[i];
+            if (isShortOptionToken(arg)) {
+                i += replaceTaggedOption(SHORT_OPTION_PREFIX, arg.substring(SHORT_OPTION_PREFIX.length()), args, finalArgs, i);
+            } else if (isLongOptionToken(arg)) {
+                i += replaceTaggedOption(LONG_OPTION_PREFIX, arg.substring(LONG_OPTION_PREFIX.length()), args, finalArgs, i);
+            } else {
+                // Positional arg, etc., just add it
+                finalArgs.add(arg);
+                i++;
+            }
+        }
+        return finalArgs.toArray(new String[finalArgs.size()]);
+    };
+
+    /**
+     * Process a single option and add it to the curated args list. If the option is tagged, add the
+     * curated key and value. Otherwise just add the raw option.
+     *
+     * @param optionPrefix the actual option prefix used for this option, either "-" or "--"
+     * @param optionString the option string including everything but the prefix
+     * @param userArgList the raw arguments provided by the user, used to retrieve the value corresponding
+     *                    to this option
+     * @param finalArgList the curated argument list
+     * @param i the index into the raw arg list of the option we're currently processing
+     * @return the number of items on the user arg list consumed by this method. This will either 1 if this
+     * option is not tagged) or 2 (if the option is tagged)
```

Comment:
Would be cleaner if this took an `Iterator<String>` for the user args and returned a `List<String>` tokens to be added to the final args. That way you don't need to pass around indices and pass in lists to be modified.